src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonObject.KeyCollection.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Text.Json.Node
+{
+    public partial class JsonObject
+    {
+        private KeyCollection? _keyCollection;
+
+        private KeyCollection GetKeyCollection(JsonObject jsonObject)
+        {
+            CreateList();
+            return _keyCollection ??= new KeyCollection(jsonObject);
+        }
+
+        private sealed class KeyCollection : ICollection<string>
```

Comment:
Per practice in most System.Collections types both this and `ValueCollection` could be changed into public structs.

Reply:
These are for to support the IDictionary contract for
`ICollection<string> IDictionary<string, JsonNode?>.Keys` and
`ICollection<JsonNode?> IDictionary<string, JsonNode?>.Values`.

If we only used `Dictionary<>` as the internal implementation then we would just delegate to the dictionary's implementation of these methods instead of adding these `KeyCollection` and `ValueCollection` classes. However, since we use a list until the threshold is hit, these extra types are just used for that and thus are hidden.